algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs

```diff
@@ -15,360 +15,191 @@
-                // \sum_i{c_i*(H_i.size()/H.size())*(lhs_i/v_H_i - x_g_1_i*v_H/v_H_i)} + (m/v_H - x_g_1_m) = \sum{h_1_i*v_H}
-                // \sum_i{c_i*(H_i.size()/H.size())*(lhs_i/v_H_i} + m/v_H = \sum{h_1_i*v_H} + \sum{c_i*x_g_1_i*(v_H/v_H_i)*(H_i.size()/H.size())} + x_g_1_m
-                // (\sum_i{c_i*s_i*lhs_i} + m)/v_H = \sum{h_1_i*v_H} + \sum{c_i*s_i*x_g_1_i} + x_g_1_m
-                // (\sum_i{c_i*s_i*lhs_i} + m)/v_H = h_1*v_H + x_g_1
-                // That's what we're computing here.
-
-                let sumcheck_time = start_timer!(|| format!("Compute sumcheck h and g for {}", _circuit_id));
-                let multiplier = circuit_combiner * constraint_domain.size_as_field_element * max_constraint_domain_inv;
-                cfg_iter_mut!(circuit_specific_lhs.coeffs).for_each(|c| *c *= multiplier);
-                let (h_1_i, mut xg_1_i) = circuit_specific_lhs.divide_by_vanishing_poly(constraint_domain).unwrap();
-                xg_1_i = xg_1_i.mul_by_vanishing_poly(max_constraint_domain);
-                let (xg_1_i, remainder) = xg_1_i.divide_by_vanishing_poly(constraint_domain).unwrap();
-                assert!(remainder.is_zero());
-                end_timer!(sumcheck_time);
-
-                (h_1_i, xg_1_i)
-            });
-        }
-
-        let mut h_1_sum = DensePolynomial::zero();
-        let mut xg_1_sum = DensePolynomial::zero();
-        for (mut h_1_i, mut xg_1_i) in job_pool.execute_all().into_iter() {
-            h_1_sum += &core::mem::take(&mut h_1_i);
-            xg_1_sum += &core::mem::take(&mut xg_1_i);
+            for (j, (instance_combiner, z_a, z_b, z_c)) in
+                itertools::izip!(instance_combiners, z_a, z_b, z_c).enumerate()
+            {
+                job_pool.add_job(move || {
+                    let mut instance_lhs = DensePolynomial::zero();
+                    let za_label = witness_label(circuit.id, "z_a", j);
+                    let zb_label = witness_label(circuit.id, "z_b", j);
+                    let zc_label = witness_label(circuit.id, "z_c", j);
+                    let z_a = Self::calc_z_m(za_label, z_a, constraint_domain, circuit, None);
+                    let z_b = Self::calc_z_m(zb_label, z_b, constraint_domain, circuit, None);
+                    let z_c = Self::calc_z_m(zc_label, z_c, constraint_domain, circuit, None);
+                    let mut multiplier_2 = PolyMultiplier::new();
+                    multiplier_2.add_precomputation(fft_precomputation, ifft_precomputation);
+                    multiplier_2.add_polynomial(z_a, "z_a");
+                    multiplier_2.add_polynomial(z_b, "z_b");
+                    let mut rowcheck = multiplier_2.multiply().unwrap();
+                    cfg_iter_mut!(rowcheck.coeffs).zip(&z_c.coeffs).for_each(|(ab, c)| *ab -= c);
+
+                    if !rowcheck.coeffs.iter().filter(|&c| *c != F::zero()).collect_vec().is_empty() {
```

Comment:
```suggestion
                    if !rowcheck.coeffs.iter().filter(|&c| *c != F::zero()).next().is_none() {
```
avoids an allocation

Reply:
febb1da7e